com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/otaserver/ZclOtaUpgradeServer.java

```diff
@@ -451,12 +451,11 @@ public void run() {
                     }
 
                     updateStatus(ZigBeeOtaServerStatus.OTA_UPGRADE_FIRMWARE_RESTARTING);
-                    CommandResult response = cluster.sendCommand(new UpgradeEndResponse(
-                            otaFile.getManufacturerCode(),
-                            otaFile.getImageType(),
-                            otaFile.getFileVersion(),
-                            0,
-                            0)).get();
+                    UpgradeEndResponse upgradeEndResponse = new UpgradeEndResponse(otaFile.getManufacturerCode(),
+                            otaFile.getImageType(), otaFile.getFileVersion(), 0, 0);
+                    upgradeEndResponse.setDisableDefaultResponse(true);
+
+                    CommandResult response = cluster.sendCommand(upgradeEndResponse).get();
```

Comment:
Can I suggest that instead of the above, use `cluster.sendResponse`? This is probably better and will automatically disable the default response.

Reply:
Thanks for the review.

I just tried to change this, but this seems to become a larger change unfortunately.
The method `cluster.sendResponse(command, response);`  requires the `command` to which we are responding.

We receive the `command` in `private boolean handleUpgradeEndCommand(UpgradeEndCommand command) {`, but
where the response to it should be send, we are in a separate thread within the `completeUpgrade()` method.

So I would somehow need to store the received command and later read it from within the `completeUpgrade` thread. Also I have to make it a member variable of the `ZclOtaUpgradeServer`, something like `receivedEndCommand` or so.

Am I overlooking something here? If not, is it worth the effort, or shall we continue with the 2 line change I have already provided?

Reply:
@cdjackson Once you find some time, I would appreciate if you could have a look at my reply. Thanks.

Reply:
Sorry for the slow response @triller-telekom. So I guess we have two options...

1. As you originally proposed - ie just setting the disable response flag
2. We could modify the `completeUpgrade` method to be `completeUpgrade(UpgradeEndCommand command)`.

If we do the second one, I think this works ok and we just change line 454 to -:

```
                   cluster.sendResponse(command, new UpgradeEndResponse(
                            otaFile.getManufacturerCode(),
                            otaFile.getImageType(),
                            otaFile.getFileVersion(),
                            0,
                            0));
```

The following check for the timeout then doesn't work since the `sendResponse` method has no return. This is a bit of a downside in this respect as it would be kind of useful to know if it is received even if we're not expecting the response.

I'm a fan of using the `sendResponse` where possible so that stuff is handled in a consistent manner. Maybe `sendResponse` ought to return the `Future` so that callers can still tell if there was an `ACK`...

What do you think?

Reply:
I just had a more indepth look at the second option. It needs a couple of small changes to the autocoder and the `ZclCluster` class so that `sendResponse` method becomes -:

```
    protected Future<CommandResult> sendResponse(ZclCommand command, ZclCommand response) {
        response.setDestinationAddress(command.getSourceAddress());
        response.setCommandDirection(command.getCommandDirection().getResponseDirection());
        response.setTransactionId(command.getTransactionId());
        response.setDisableDefaultResponse(true);

        if (isManufacturerSpecific()) {
            response.setManufacturerCode(getManufacturerCode());
        }

        response.setApsSecurity(apsSecurityRequired);

        return sendCommand(response);
    }
```

and in `ZclOtaUpgradeCluster` -:

```
    public Future<CommandResult> sendResponse(ZclOtaUpgradeCommand command, ZclOtaUpgradeCommand response) {
        return super.sendResponse(command, response);
    }
```

This would ultimately be resolved by a change to the autocoder, but I think these changes should be enough to test with if you think this is a good approach...

Reply:
Thanks for your suggestions, we will give the changes a try and see if more changes would be necessary or if we would even have to break the API. If that s not the case, I think we could do it as you suggested.

Reply:
@cdjackson 
@triller-telekom and I are working on the implementation of the changes that you suggested to realize option 2. Unfortunately we run into a problem when changing `completeUpgrade()` to `completeUpgrade(UpgradeEndCommand command)`. The problem is that `ZigBeeConsoleOtaUpgradeCommand.cmdOtaComplete(...)` invokes `ZclOtaUpgradeServer.completeUpgrade()` and no UpgradeEndCommand is available there in order to call `completeUpgrade(UpgradeEndCommand command)`. Is there anything we can do about that?

Reply:
Thanks @TomTravaglino - I missed this call when I looked at this the other day...

So I guess this shows that actually there are two ways that the OTA can be completed - either in response to the `UpgradeEndCommand`, or as a command at any other time. If I remember well, it can be a command (rather than a response) if the server wanted to wait till a later time to actually perform the upgrade, so I think this is a valid use case as per the spec.

So, maybe we should have both options available - `completeUpgrade()` and ` completeUpgrade(UpgradeEndCommand command)` - one sending as a command, and one as a response? `completeUpgrade()` could just be a proxy for `completeUpgrade(null)` and then in the ` completeUpgrade(UpgradeEndCommand command)` method, it checks for null and sends appropriately...

What do you and @triller-telekom think?

Reply:
Thanks for your suggestions Chris.

I also think that we should offer both versions, so we should change the `completeUpgrade()` as you sketched to have both signatures available.

@TomTravaglino This sounds like a good solution, right?

Reply:
@cdjackson @triller-telekom 
Sorry for not replying (very busy yesterday). Well, I think it's a good solution that covers both cases.

Reply:
@cdjackson Thank you for your input. Tommaso has integrated your suggestions and I have pushed it onto my PR.

Could you please have another look?

Reply:
Thanks @TomTravaglino and @triller-telekom - this looks fine to me. I've not actually run the code, but it looks good and in-line with what I was thinking so if you've tested the OTA completion, then I'm happy with this.

Reply:
@cdjackson 
We successfully tested the new solution. The update completed without errors. Here is a quick comparison old vs new

Old solution: 
- In `UpgradeEndResponse` the parameter `disableDefaultResponse` is false
- We receive a `DefaultResponse` from the device
- OTA fails
```
2021-02-26T11:47:28.325+0100   432  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - TX CMD: UpgradeEndResponse [Ota Upgrade: 0000/0 -> 4081/1, cluster=0019, TID=E7, manufacturerCode=4209, imageType=235, fileVersion=440, currentTime=0, upgradeTime=0]
2021-02-26T11:47:28.328+0100   432  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - TX ZCL: ZclHeader [frameType=CLUSTER_SPECIFIC_COMMAND, manufacturerSpecific=false, direction=SERVER_TO_CLIENT, disableDefaultResponse=false, manufacturerCode=0, sequenceNumber=231, commandId=7]
...
2021-02-26T11:47:47.679+0100   432  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - RX CMD: DefaultResponse [Ota Upgrade: 4081/1 -> 0000/1, cluster=0019, TID=E7, commandIdentifier=7, statusCode=FAILURE]
2021-02-26T11:47:47.680+0100   432  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.transaction.ZigBeeTransactionManager] - notifyTransactionCommand: DefaultResponse [Ota Upgrade: 4081/1 -> 0000/1, cluster=0019, TID=E7, commandIdentifier=7, statusCode=FAILURE] 
2021-02-26T11:47:47.681+0100   432  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.otaserver.ZclOtaUpgradeServer] - 4081/1 OTA status updated to OTA_UPGRADE_FAILED.
2021-02-26T11:47:47.743+0100   432  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.otaserver.ZclOtaUpgradeServer] - 4081/1 OTA status updated to OTA_CANCELLED.
``` 

New solution:
- In `UpgradeEndResponse` the parameter `disableDefaultResponse` is true
- No `DefaultResponse` from the device
- OTA succeeds

```
2021-05-04T15:24:04.372+0200   472  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.otaserver.ZclOtaUpgradeServer] - 9282/1 OTA status updated to OTA_TRANSFER_COMPLETE.
...
2021-05-04T15:24:05.182+0200   472  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - TX CMD: UpgradeEndResponse [Ota Upgrade: 0000/0 -> 9282/1, cluster=0019, TID=46, manufacturerCode=4209, imageType=235, fileVersion=440, currentTime=0, upgradeTime=0]
2021-05-04T15:24:05.183+0200   472  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.ZigBeeNetworkManager] - TX ZCL: ZclHeader [frameType=CLUSTER_SPECIFIC_COMMAND, manufacturerSpecific=false, direction=SERVER_TO_CLIENT, disableDefaultResponse=true, manufacturerCode=0, sequenceNumber=70, commandId=7]
...
2021-05-04T15:24:25.062+0200   472  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.otaserver.ZclOtaUpgradeServer] - 9282/1 : OTA file version to be installed=440 Received ReadAttributesResponse: ReadAttributesResponse [Ota Upgrade: 9282/1 -> 0000/1, cluster=0019, TID=0F, records=[ReadAttributeStatusRecord [status=SUCCESS, attributeIdentifier=2, attributeDataType=UNSIGNED_32_BIT_INTEGER, attributeValue=440]]]
2021-05-04T15:24:25.063+0200   472  DEBUG          [symbolicName=com.zsmartsystems.zigbee, className=com.zsmartsystems.zigbee.app.otaserver.ZclOtaUpgradeServer] - 9282/1 OTA status updated to OTA_UPGRADE_COMPLETE.

``` 

Attached you can find the logs of a test with the old solution and a test with the new solution.
[logs-old.log](https://github.com/zsmartsystems/com.zsmartsystems.zigbee/files/6425515/logs-old.log)
[logs-new.log](https://github.com/zsmartsystems/com.zsmartsystems.zigbee/files/6425516/logs-new.log)

Reply:
@cdjackson Have you already had the chance to take a look at my last post?

Reply:
Sorry for the delay @TomTravaglino - this week has been a bit stressful and busy as we have been trying to buy a house and in New Zealand there are strict deadlines on getting everything completed.

This looks good - thanks for the testing. I'll merge this. Thanks also @triller-telekom